Implemented Cache Lookup Policy for acquireTokenSilent (#5014)
AzureAD/microsoft-authentication-library-for-js: lib/msal-browser/docs/acquire-token.md

You must pass a request object to the acquireToken APIs. This object allows you to use different parameters in the request. See [here](./request-response-object.md) for more information on the request object parameters. Scopes are required for all acquireToken calls.

A Cache Lookup Policy can be optionally provided to the request. The Cache Lookup Policies are:

> - `CacheLookupPolicy.Default` - `acquireTokenSilent` will attempt to retrieve an access token from the cache. If the access token is expired or cannot be found the refresh token will be used to acquire a new one. Finally, if the refresh token is expired `acquireTokenSilent` will attempt to acquire new access and refresh tokens.

Reviewer: ```suggestion
- `CacheLookupPolicy.Default` - `acquireTokenSilent` will attempt to retrieve an access token from the cache. If the access token is expired or cannot be found the refresh token will be used to acquire a new one. Finally, if the refresh token is expired, `acquireTokenSilent` will attempt to silently acquire a new access token, id token, and refresh token.
```

- `CacheLookupPolicy.AccessTokenAndRefreshToken` - `acquireTokenSilent` will attempt to retrieve an access token from the cache. If the access token is expired or cannot be found, the refresh token will be used to acquire a new one. If the refresh token is expired, it will not be renewed and `acquireTokenSilent` will fail.
- `CacheLookupPolicy.RefreshToken` - `acquireTokenSilent` will not attempt to retrieve access tokens from the cache and will instead attempt to exchange the cached refresh token for a new access token. If the refresh token is expired, it will not be renewed and `acquireTokenSilent` will fail.
- `CacheLookupPolicy.RefreshTokenAndNetwork` - `acquireTokenSilent` will not look in the cache for the access token. It will go directly to network with the cached refresh token. If the refresh token is expired an attempt will be made to renew it. This is equivalent to setting `forceRefresh: true`.
- `CacheLookupPolicy.Skip` - `acquireTokenSilent` will attempt to renew both access and refresh tokens. It will not look in the cache. This will always fail if 3rd party cookies are blocked by the browser.

- Popup

```javascript
var request = {
    scopes: ["Mail.Read"],

>     cacheLookupPolicy: 0 // Default

Reviewer 2: I suggest using the enum in the doc since we want to promote readable code.
```suggestion
    cacheLookupPolicy: CacheLookupPolicy.Default
```

You must pass a request object to the acquireToken APIs. This object allows you to use different parameters in the request. See [here](./request-response-object.md) for more information on the request object parameters. Scopes are required for all acquireToken calls.

A Silent Token Retrieval Strategy can be optionally provided to the request. The Retrieval Strategies are:

> - CacheAndNetwork - acquireTokenSilent will first look in the cache before going to the network with the existing refresh token. The refresh token will be renewed if it is expired. This is existing default behavior.

Reviewer 3: ```suggestion
- Default - `acquireTokenSilent` will attempt to retrieve an access token from the cache. If the access token is expired or cannot be found the refresh token will be used to acquire a new one. Finally, if the refresh token is expired `acquireTokenSilent` will attempt to acquire new access and refresh tokens.
```

> - Default - `acquireTokenSilent` will attempt to retrieve an access token from the cache. If the access token is expired or cannot be found the refresh token will be used to acquire a new one. Finally, if the refresh token is expired `acquireTokenSilent` will attempt to acquire new access and refresh tokens.

Reviewer: Please show entire enum value and wrap in backticks.

```suggestion
- `CacheLookupPolicy.Default` - `acquireTokenSilent` will attempt to retrieve an access token from the cache. If the access token is expired or cannot be found the refresh token will be used to acquire a new one. Finally, if the refresh token is expired `acquireTokenSilent` will attempt to acquire new access and refresh tokens.
```

Reviewer: Also, please show a code snippet demonstrating basic usage, thanks!

> - CacheOnly - acquireTokenSilent will only look in the cache, and will not go to the network with the existing refresh token. The refresh token will not be renewed if it is expired.

Reviewer 3: ```suggestion
- CacheOnly - `acquireTokenSilent` will only look for access tokens in the cache. It will not attempt to renew access or refresh tokens.
```

> - CacheOrExistingRefreshToken - acquireTokenSilent will first look in the cache before going to the network with the existing refresh token. The refresh token will not be renewed if it is expired.

Reviewer 3: ```suggestion
- CacheOrRefreshToken - `acquireTokenSilent` will attempt to retrieve an access token from the cache. If the access token is expired or cannot be found, the refresh token will be used to acquire a new one. If the refresh token is expired, it will not be renewed and `acquireTokenSilent` will fail.
```

> - NetworkWithExistingRefreshTokenOnly - acquireTokenSilent will only go to network with the existing refresh token, and will not look in the cache. The refresh token will not be renewed if it is expired.

Reviewer 3: ```suggestion
- RefreshTokenOnly - `acquireTokenSilent` will not attempt to retrieve access tokens from the cache and will instead attempt to exchange the cached refresh token for a new access token. If the refresh token is expired, it will not be renewed and `acquireTokenSilent will fail.
```

> - NetworkWithRefreshToken - acquireTokenSilent will only go to network with the existing refresh token, and will not look in the cache. The refresh token not be renewed if it is expired. This is existing "`forceRefresh: true`" behavior.

Reviewer 3: ```suggestion
- ForceRefresh - `acquireTokenSilent` will not attempt to retrieve access tokens from the cache and will instead attempt to exchange the cached refresh token for a new access token. if the refresh token is expired `acquireTokenSilent` will attempt to acquire new access and refresh tokens. (This is equivalent to setting `forceRefresh: true` on the request)
```

Reviewer: Hmm don't like this as much, as it isn't consistent with the others, and I don't really like "force refresh"

Reviewer 3: `RefreshTokenOrRenewAll`? Which makes the ssoSilent-like behavior `RenewAll`?

Author: @jasonnutter thoughts about Thomas's comment?

Reviewer: And then `NetworkWithExistingRefreshTokenOnly` -> `RefreshTokenOnly`? Would like some consistency between the network only options.

Reviewer 4: In that case shouldn't `CacheOrExistingRefreshToken` be renamed `CacheOrNetworkWithExistingRefreshTokenOnly`?

Reviewer 4: My two cents:  Naming consistency across, `cache`, `networkWithExistingRT` and `networkWithRT` needed. Seems inconsistent and confusing as it stands. It is okay to add comments relating to `default`, `forceRefresh` or `ssoSilent` in the description, however I prefer the naming convention of the options to be uniform. And @Robbie-Microsoft may be a table in docs will help.

Mapping them to existing APIs and what is new makes it more readable IMO for usage docs.

Reviewer 3: @jasonnutter I'm good with that, I think that's what I had suggested above. @Sameerag `CacheOrExistingRefreshToken` became `CacheOrRefreshToken`

> - NetworkOnly - acquireTokenSilent will only go to network, and not look in the cache. It will not use the existing refresh token, regardless of whether it is expired or not.

Reviewer 3: IMO this option can be left off as this is the `ssoSilent` behavior

Reviewer: True, but I'd like to keep the semantics separate and not have apps call ssosilent when they could call acquiretokensilent.

Reviewer 3: That's fair. FWIW I do think this is the least likely option to be used and if we keep it we should discourage its use as much as possible.

Reviewer 4: Can we refactor the code to call `ssoSilent` when this option is selected for consistency?

Reviewer: It already is (effectively, calls the same internal apis).

Reviewer: I don't think we should call ssosilent directly because ssosilent is meant as a public api and needs different semantics than an internal api. Also, it would make metrics and telemetry confusing.

> - RefreshTokenAndNetwork - acquireTokenSilent will only go to network with the existing refresh token, and will not look in the cache. The refresh token will be renewed if it is expired. This is existing "`forceRefresh: true`" behavior.

Reviewer 3: ```suggestion
- RefreshTokenAndNetwork - acquireTokenSilent will not look in the cache for the access token. It will directly go to network with the cached refresh token. If the refresh token is expired an attempt will be made to renew it. This is equivalent to setting `forceRefresh: true`.
```

> - Skip - `acquireTokenSilent` will only go to network and not look in the cache. It will not use the existing refresh token, regardless of whether it is expired or not.

Reviewer 3: ```suggestion
- Skip - `acquireTokenSilent` will attempt to renew both access and refresh tokens. It will not look in the cache. This will always fail if 3rd party cookies are blocked by the browser.
```

> - NetworkOnly - `acquireTokenSilent` will only go to network and not look in the cache. It will not use the existing refresh token, regardless of whether it is expired or not.

Reviewer 3: Thought we changed this to `RefreshTokenOrRenewAll` and `RenewAll`? Don't particularly like the word "network" in this context given all options other than CacheOnly use the network.

Author: I chatted with @jasonnutter, and we decided to stick with what it currently is, which is what we came up with in the design review.

```javascript
var request = {
    scopes: ["Mail.Read"],

>     cacheLookupPolicy: 0 // Default

Reviewer 2: Same here.
